src/mpid/ch4/am/rndv/rndv.c

```diff
@@ -51,7 +57,140 @@ int MPIDI_am_rndv_init(void)
+        MPIDIG_REQUEST(rreq, context_id) = rts_hdr->context_id;
+        MPIDIG_REQUEST(rreq, req->status) |= MPIDIG_REQ_IN_PROGRESS;
 
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+        MPIDI_REQUEST(rreq, is_local) = rts_hdr->is_local;
+#endif
+        MPID_THREAD_CS_ENTER(VCI, MPIDIU_THREAD_MPIDIG_GLOBAL_MUTEX);
+        if (root_comm) {
+            MPIR_Comm_add_ref(root_comm);
+            MPIDIG_enqueue_unexp(rreq, &MPIDIG_COMM(root_comm, unexp_list));
+        } else {
+            MPIR_Comm *root_comm_again;
+            /* This branch means that last time we checked, there was no communicator
+             * associated with the arriving message.
+             * In a multi-threaded environment, it is possible that the communicator
+             * has been created since we checked root_comm last time.
+             * If that is the case, the new message must be put into a queue in
+             * the new communicator. Otherwise that message will be lost forever.
+             * Here that strategy is to query root_comm again, and if found,
+             * simply re-execute the per-communicator enqueue logic above. */
+            root_comm_again = MPIDIG_context_id_to_comm(rts_hdr->context_id);
+            if (unlikely(root_comm_again != NULL)) {
+                MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_MPIDIG_GLOBAL_MUTEX);
+                MPL_free(MPIDIG_REQUEST(rreq, buffer));
+                MPIR_Request_free(rreq);
+                MPID_Request_complete(rreq);
+                rreq = NULL;
+                root_comm = root_comm_again;
+                goto root_comm_retry;
+            }
+            MPIDIG_enqueue_unexp(rreq,
+                                 MPIDIG_context_id_to_uelist(MPIDIG_REQUEST(rreq, context_id)));
+        }
+        MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_MPIDIG_GLOBAL_MUTEX);
+    } else {
+        /* Matching receive was posted, tell the netmod */
+        MPIR_Comm_release(root_comm);   /* -1 for posted_list */
+        MPIDIG_REQUEST(rreq, req->status) |= MPIDIG_REQ_LONG_RTS;
+        MPIDIG_REQUEST(rreq, req->rreq.peer_req_ptr) = rts_hdr->sreq_ptr;
+        MPIDIG_REQUEST(rreq, rank) = rts_hdr->src_rank;
+        MPIDIG_REQUEST(rreq, tag) = rts_hdr->tag;
+        MPIDIG_REQUEST(rreq, context_id) = rts_hdr->context_id;
+        MPIDIG_REQUEST(rreq, req->status) |= MPIDIG_REQ_IN_PROGRESS;
+        /* Mark `match_req` as NULL so that we know nothing else to complete when
+         * `unexp_req` finally completes. (See MPIDI_recv_target_cmpl_cb) */
+        MPIDIG_REQUEST(rreq, req->rreq.match_req) = NULL;
+
+        /* prepare clear to send header */
+        hdr.handler_id = cts_hnd;
+        cts_hdr->sreq_ptr = rts_hdr->sreq_ptr;
+        cts_hdr->rreq_ptr = rreq;
+
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+        if (MPIDI_REQUEST(rreq, is_local)) {
+            mpi_errno =
+                MPIDI_SHM_am_send_hdr(rts_hdr->src_rank, sizeof(MPIDI_am_rndv_hdr_t), &cts_hdr,
```

Comment:
@hzhou can you remind me why we need `handler_id` and `completion_id` in the `hdr_send`? Should not these be handled by the upper layer rather than the NM or SHM (either when the send returns or in the progress engine)? I am momentarily not passing them down as the AM header already contains the `handler_id`. Please look at the `rndv_pre.h` file to see how the AM header would look like?

Reply:
When netmod receives a active message, it need know its handler_id in order to dispatch. The protocol header (I assume which you refer as AM header) is defined by protocol layer and is opaque to netmod. If you put handler_id in protocol header and omit from function parameter, then you are asking netmod to peek into the protocol header, which means the protocol header is no longer opaque.

Handler_id is really only of interest to transport/netmod layer so it doesn't make sense to pack into protocol header. But that mean the transport layer will have to add one additional header -- admittedly it is a bit cumbersome. What @raffenet and I came up is to simply require all protocol header reserve a padding space at the front of the header. This reserved space is for transport layer to fill, with handler_id and other fields. We leave this reserved header space opaque to upper layer so the different transport layer can define their own structure as fit. See comments in https://github.com/pmodels/mpich/pull/4236/commits/0e2a92ca472031ffeca30edba66427784df5c371.

Reply:
No `MPIDI_am_hdr_t` is equivalent to `MPIDIG_hdr_t`. Only this contains the `handler_id` so the netmod does not need to peek into the protocol header. Additionally the am_hdr also has additional space for the netmod to add its information. This is based on my memory of the discussion we had in which having the netmod to add information by creating its own header and copying am_hdr and protocol header into was wasteful of memory copies.